RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitProducers.cc

```diff
@@ -0,0 +1,164 @@
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include <iomanip>  // for std::setw
+#include <future>
+
+// Alpaka imports
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/stream/EDProducer.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/EDPutToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/ESGetToken.h"
+#include "HeterogeneousCore/AlpakaCore/interface/alpaka/Event.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/CopyToDevice.h"
+
+// includes for data formats
+#include "DataFormats/HGCalDigi/interface/HGCalDigiHost.h"
+#include "DataFormats/HGCalDigi/interface/alpaka/HGCalDigiDevice.h"
+#include "DataFormats/HGCalRecHit/interface/HGCalRecHitHost.h"
+#include "DataFormats/HGCalRecHit/interface/alpaka/HGCalRecHitDevice.h"
+
+// includes for size, calibration, and configuration parameters
+#include "CondFormats/DataRecord/interface/HGCalElectronicsMappingRcd.h"
+#include "CondFormats/DataRecord/interface/HGCalModuleConfigurationRcd.h"
+#include "CondFormats/HGCalObjects/interface/HGCalMappingModuleIndexer.h"
+#include "CondFormats/HGCalObjects/interface/HGCalCalibrationParameterHost.h"
+#include "CondFormats/HGCalObjects/interface/alpaka/HGCalCalibrationParameterDevice.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/alpaka/HGCalRecHitCalibrationAlgorithms.h"
+
+template <class T>
+double duration(T t0, T t1) {
+  auto elapsed_secs = t1 - t0;
+  typedef std::chrono::duration<float> float_seconds;
+  auto secs = std::chrono::duration_cast<float_seconds>(elapsed_secs);
+  return secs.count();
+}
+
+inline std::chrono::time_point<std::chrono::steady_clock> now() { return std::chrono::steady_clock::now(); }
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  using namespace cms::alpakatools;
+
+  class HGCalRecHitsProducer : public stream::EDProducer<> {
+  public:
+    explicit HGCalRecHitsProducer(const edm::ParameterSet&);
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    void produce(device::Event&, device::EventSetup const&) override;
+    void beginRun(edm::Run const&, edm::EventSetup const&) override;
+    edm::ESWatcher<HGCalElectronicsMappingRcd> calibWatcher_;
+    edm::ESWatcher<HGCalModuleConfigurationRcd> configWatcher_;
+    const edm::EDGetTokenT<hgcaldigi::HGCalDigiHost> digisToken_;
+    device::ESGetToken<hgcalrechit::HGCalCalibParamDevice, HGCalModuleConfigurationRcd> calibToken_;
+    device::ESGetToken<hgcalrechit::HGCalConfigParamDevice, HGCalModuleConfigurationRcd> configToken_;
+    const device::EDPutToken<hgcalrechit::HGCalRecHitDevice> recHitsToken_;
+    HGCalRecHitCalibrationAlgorithms calibrator_;  // cannot be "const" because the calibrate() method is not const
```

Comment:
On a quick look the `calibrate()` method could be easily made `const`, or did I miss something?

Reply:
You're right, thanks for insisting. When the test was made we made a mistake with the placement of the const declaration and got erroneous results. This is now solved.